File: leetCode/3sum_15.cpp

#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());        

        for (int i = 0; i < nums.size(); i++)
        {
            int a = nums[i];
            if (i > 0 && a == nums[i-1])
                continue;
            int left = i + 1 , right = nums.size() - 1;
            while(left < right) {
                int sum = a + nums[left] + nums[right];
                if (sum > 0) {
                    right--;
                } 
                else if (sum < 0) {
                    left++;  
                }
                // for sum = 0
                else {
                    ans.push_back({a, nums[left], nums[right]});
                    left++;
                    while (nums[left] == nums[left - 1] && left < right) {
                        left++;
                    }
                }
            }
        }

        return ans;
    }
};

int main()
{
    Solution solution;

    vector<vector<int>> testCases = {
        {-1, 0, 1, 2, -1, -4},
        {0, 0, 0},
        {1, -1, -1, 0},
        {-2, 0, 1, 1, 2}};

    for (auto &nums : testCases)
    {
        cout << "Input: [";
        for (int i = 0; i < nums.size(); i++)
        {
            cout << nums[i];
            if (i < nums.size() - 1)
                cout << ",";
        }
        cout << "]\nOutput: ";

        vector<vector<int>> result = solution.threeSum(nums);
        cout << "[";
        for (int i = 0; i < result.size(); i++)
        {
            cout << "[";
            for (int j = 0; j < result[i].size(); j++)
            {
                cout << result[i][j];
                if (j < result[i].size() - 1)
                    cout << ",";
            }
            cout << "]";
            if (i < result.size() - 1)
                cout << ",";
        }
        cout << "]\n\n";
    }
    return 0;
}

File: leetCode/3Sum_15.py

from typing import List

class Solution:
    def threeSum(self , nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        for i , a in enumerate(nums):
            if i > 0  and a == nums[i - 1]:
                continue
            
            left , right = i+1 , len(nums) - 1
            while left < right:
                threeSum = a + nums[left] + nums[right]
                if threeSum > 0:
                    right -= 1
                elif threeSum < 0:
                    left += 1
                else:
                    res.append([a, nums[left], nums[right]])
                    left += 1
                    while nums[left] == nums[left - 1] and left < right:
                        left += 1
        return res
                    

if __name__ == "__main__":
    solution = Solution()
    test_cases = [
        [-1, 0, 1, 2, -1, -4],
        [0, 0, 0],
        [1, -1, -1, 0]
    ]
    
    for nums in test_cases:
        print(f"Input: {nums}")
        print(f"Output: {solution.threeSum(nums)}\n")

    
